pkg/http/server.go

```diff
@@ -0,0 +1,184 @@
+type Server struct {
+	http.Server
+	Addrs                  []string                            // addresses on which the server listen for new connection.
+	ShutdownTimeout        time.Duration                       // timeout used for graceful server shutdown.
+	TCPKeepAliveTimeout    time.Duration                       // timeout used for underneath TCP connection.
+	UpdateBytesReadFunc    func(int)                           // function to be called to update bytes read in bufConn.
+	UpdateBytesWrittenFunc func(int)                           // function to be called to update bytes written in bufConn.
+	ErrorLogFunc           func(error, string, ...interface{}) // function to be called on errors.
+	listenerMutex          *sync.Mutex                         // to guard 'listener' field.
+	listener               *httpListener                       // HTTP listener for all 'Addrs' field.
+	inShutdownMutex        *sync.Mutex                         // to guard 'inShutdown' field
+	inShutdown             bool                                // indicates whether the server is in shutdown or not
+	requestCount           int32                               // counter holds no. of request in process.
+}
+
+// Start - start HTTP server
+func (srv *Server) Start() (err error) {
+	// Take a copy of server fields.
+	tlsConfig := srv.TLSConfig
+	readTimeout := srv.ReadTimeout
+	writeTimeout := srv.WriteTimeout
+	handler := srv.Handler
+
+	addrs := srv.Addrs
+	tcpKeepAliveTimeout := srv.TCPKeepAliveTimeout
+	updateBytesReadFunc := srv.UpdateBytesReadFunc
+	updateBytesWrittenFunc := srv.UpdateBytesWrittenFunc
+	errorLogFunc := srv.ErrorLogFunc
+
+	// Create new HTTP listener.
+	var listener *httpListener
+	listener, err = newHTTPListener(
+		addrs,
+		tlsConfig,
+		tcpKeepAliveTimeout,
+		readTimeout,
+		writeTimeout,
+		updateBytesReadFunc,
+		updateBytesWrittenFunc,
+		errorLogFunc,
+	)
+	if err != nil {
+		return err
+	}
+
+	// Wrap given handler to do additional
+	// * return 503 (service unavailable) if the server in shutdown.
+	wrappedHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		atomic.AddInt32(&srv.requestCount, 1)
+		defer atomic.AddInt32(&srv.requestCount, -1)
+
+		// If server is in shutdown, return 503 (service unavailable)
+		srv.inShutdownMutex.Lock()
+		inShutdown := srv.inShutdown
+		srv.inShutdownMutex.Unlock()
+		if inShutdown {
+			w.WriteHeader(http.StatusServiceUnavailable)
+			return
+		}
+
+		// Handle request using passed handler.
+		handler.ServeHTTP(w, r)
+	})
+
+	srv.listenerMutex.Lock()
+	srv.Handler = wrappedHandler
+	srv.listener = listener
+	srv.listenerMutex.Unlock()
+
+	// Start servicing with listener.
+	return srv.Server.Serve(listener)
+}
+
+// Shutdown - shuts down HTTP server.
+func (srv *Server) Shutdown() error {
+	srv.inShutdownMutex.Lock()
+	if !srv.inShutdown {
+		// Indicate the server is in shutdown.
+		srv.inShutdown = true
```

Comment:
The approach is different here.  Its not only guard `src.inShutdown`, it also guards shutdown operation ie there is no need to have one more mutex with `sync.Atomic`